projects/core/src/product/services/product-page-meta.resolver.ts

```diff
@@ -179,4 +181,32 @@ export class ProductPageMetaResolver
   resolveRobots(): Observable<PageRobotsMeta[]> {
     return of([PageRobotsMeta.FOLLOW, PageRobotsMeta.INDEX]);
   }
+
+  /**
+   * Resolves the canonical url for the product page using the default canonical url
+   * configuration.
+   *
+   * In case of a variant product, the baseProduct code is used to resolve the url. It's important
+   * to know that this has a few limitations:
+   * - We're not always able to get the super baseProduct, in case of multi-level variants.
+   *   OCC only exposes the direct baseProduct, which might still not resolve in the correct
+   *   canonical URL. This is business driven and subject to change in a customization.
+   * - The url resolved for the variant doesn't contain any content other then the product code.
+   *   This means that we do not provide any product data to resolve pretty URLs (for example
+   *   the product title).
+   */
+  resolveCanonicalUrl(): Observable<string> {
+    return this.product$.pipe(
+      switchMap((product) => {
+        if (product.baseProduct) {
+          const url = this.routingService.getUrl({
+            cxRoute: 'product',
+            params: { code: product.baseProduct },
+          });
+          return this.basePageMetaResolver?.resolveCanonicalUrl(undefined, url);
```

Comment:
[medium] options and url params has been recently swapped in this PR. It won't compile.
```suggestion
          return this.basePageMetaResolver?.resolveCanonicalUrl(url);
```